### PR TITLE
[CSCRIPT][WSCRIPT][BOOTDATA] Basic .wsf support

### PR DESCRIPTION
**Pull request by whindsaks**
Adds support for .wsf files with a single script block:

```
<job>
<script language="JScript">
WScript.Echo("Hello from JS");
</script>
</job>
```

.wsf files are supposed to support multiple different languages in a job but I have not been able to figure out how a single IActiveScript can parse different languages.

Notes:
- All the Windows Scripting Host icons are missing, I used one from shell32 that fits.
- The Open2 string is not translated, the translation is supposed to come from wshext.dll but ROS does not have that file yet.
- `|` without space is the style in that file.
- 0x800400C0 is reported by an error dialog on Windows, I don't know its name.
- ; WSH and the double new-line are on purpose.

### CODE REVIEW

**Comment by HBelusca** [comment]
It would be of interest for you to try to find a way to send these patches to Wine as well, so that we don't end up again with forked code for something we don't otherwise maintain.

**Comment by HBelusca** [comment]
Does Wine or sthg have a unit test for that? If not, it would be worth adding one. Already in order for the PR reviewers to know the PR works as advertised, and also, to be able to make it acceptable for sending it to Wine.

**Review by JoachimHenze — DISMISSED** [review]

File: base/applications/cmdutils/wscript/CMakeLists.txt
Comment (whindsaks) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: I don't see the problem of importing directly. This is a scripting host, no 3rd-party native code is going to throw a fit.

Windows will probably delay load the url handling. WSH was an optional component for Win95 and I just don't see the point of trying to replicate it for a reason that is no longer relevant.

As for comctl32, that was just to get the DSA functions. Replicating it on my own just to avoid some import is not worth it IMHO.

Comment (JoachimHenze) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: I disagree, we should try to match the import of our target whereever possible. Including less than MS is fine, including more is most likely not a good idea anywhere.

Comment (whindsaks) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: What good does it do if I just copy the code I need from comctl32? That just adds extra bloat and becomes a maintenance burden for zero benefit other than somehow being more NT5 compatible in ways that are not obserable.

As soon as the script uses the Shell.Application object or whatever, it's going to load shell32 and in turn shlwapi anyway.

Comment (JoachimHenze) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: Do it like MS does it please. That is the purpose of ros, and is a good thing to act by default unless there are really profound reasons to act differently. I cannot see such reasons here. And depending on whatever additional stuff is rarely a good thing to do.

Comment (JoachimHenze) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: If you want my approval for something, then it may make sense to split your tests from the implementations in a different PR. I don't have any complaints within the added tests-part.

Comment (whindsaks) [reply base/applications/cmdutils/wscript/CMakeLists.txt:16]: The wine version already links to shell32 (unlike Windows) and that pulls in shlwapi.

File: base/applications/cmdutils/cscript/CMakeLists.txt
Comment (JoachimHenze) [inline base/applications/cmdutils/cscript/CMakeLists.txt:18]: Delayed import of shlwapi might be fine (I can see that, at least transitively, being delay-loaded also on XPSP3).
Still the question for documentation purposes: Which exact call introduces the need for shlwapi.dll in your code?

The latest changeset brings me back to neutrality regarding this PR, thank you!

Comment (whindsaks) [reply base/applications/cmdutils/cscript/CMakeLists.txt:18]: > Which exact call introduces the need for shlwapi.dll in your code?

UrlCreateFromPathW

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:5]: Please fix this (you wrote this file, not the ReactOS Team)

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:160]: please remove these extra newlines here.

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:89]:
```suggestion
static void test_defaultscriptisjs(void)
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:102]:
```suggestion
static void test_simplevb(void)
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:116]:
```suggestion
static void test_defpackagejob(void)
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:135]:
```suggestion
static void test_objecttag(void)
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:118]:
```suggestion
    static LPCSTR script = ""
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:104]:
```suggestion
    static LPCSTR script = ""
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:91]:
```suggestion
    static LPCSTR script = ""
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:78]:
```suggestion
    DWORD code;
    WCHAR file[MAX_PATH];

```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:60]:
```suggestion

    wsprintfW(cmd, L"\"%s\" //nologo \"%s\"", exe, path);
```
(making the code less cluttered by separating variables and executable code)

Also, since this an APItest (and not a winetest), could you instead use `_snwprintf` or `StringCchPrintfW` (or Cb) function?

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:31]: Please use `_snwprintf` or `StringCchPrintfW` (or Cb) function.

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:18]:
```suggestion
    LRESULT e;
    DWORD size = sizeof(*out);
    *out = 0;
    e = SHGetValueA(hKey, sub, name, NULL, out, &size);
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:24]: Please make the function take the size of the path buffer in input, and have the ext as unicode.

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:139]:
```suggestion
    static LPCSTR script = ""
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [inline modules/rostests/apitests/wscript/wsf.c:31]: You should be able to use [GetTempFileName](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettempfilenamew) since you're here creating the file.

File: base/applications/cmdutils/wscript/main.c
Comment (HBelusca) [inline base/applications/cmdutils/wscript/main.c:116]:
```suggestion
static void free_globals(void)
```

File: modules/rostests/apitests/wscript/wsf.c
Comment (whindsaks) [reply modules/rostests/apitests/wscript/wsf.c:118]: No, that makes no sense.

You perhaps mean `static const char script[] = "...` but string literals are already static globals.

File: modules/rostests/apitests/wscript/wsf.c
Comment (whindsaks) [reply modules/rostests/apitests/wscript/wsf.c:60]: The string will always fit.

File: modules/rostests/apitests/wscript/wsf.c
Comment (whindsaks) [reply modules/rostests/apitests/wscript/wsf.c:31]: GetTempFileName always uses the .tmp extension but here we need a specific extension because wscript.exe uses the extension to detect the script type.

File: modules/rostests/apitests/wscript/wsf.c
Comment (HBelusca) [reply modules/rostests/apitests/wscript/wsf.c:125]: There is still a difference:
https://godbolt.org/z/jefaW6fvv
Check how assembly changes in either Func1 or Func2 if you add or remove the "static".
(and I could have suggested `static const LPCSTR script` to make the value of the pointer const and also put it in the static globals, not just the string bytes themselves.)

Comment (whindsaks) [reply modules/rostests/apitests/wscript/wsf.c:125]: I cannot believe we are going to have this discussion!

```
int foo = 1337;
SetLastError(foo);
static int bar = 42;
SetLastError(bar);
```
Are these two the same? No, of course not.

https://godbolt.org/z/TaGr4cbY8

With optimizations on, all variants turn into the same thing, the only difference is that `static LPCSTR` is basically a global variable that you are allocating for no reason.

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
@whindsaks Do you plan on sending this to Wine at some point? (This doesn't block merging this PR in any case.)

**Comment by whindsaks** [comment]
> Do you plan on sending this to Wine at some point?

It would be nice to get it fully compatible first, and by that I mean, supporting different languages in the same job like Windows. I spent a while looking at the interfaces on MSDN and I have yet to find a way to multiplex [different languages into one script context](https://en.wikipedia.org/wiki/Windows_Script_File#Mixed_language_support).

**Pull request merged**
